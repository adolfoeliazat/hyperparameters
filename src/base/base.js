import RandomState from '../utils/RandomState';

class NotImplementedError extends Error {}

export default class BaseSymbol {
  constructor(label, params) {
    this.label = label;
    this.params = params;
  }
  // eslint-disable-next-line no-unused-vars
  eval(rng) {
    // Override this method to generate a new value
    throw new NotImplementedError(this.params);
  }
}
export const expressionEval = (expr, { rng: rState }) => {
  let rng = rState;
  if (!rng) {
    rng = new RandomState();
  }
  if (expr === undefined || expr === null) {
    return expr;
  }
  if (typeof expr.eval !== 'function') {
    if (Array.isArray(expr)) {
      return expr.map(item => expressionEval(item, { rng }));
    }
    if (typeof expr === 'object') {
      return Object.keys(expr)
        .reduce((r, key) => ({ ...r, [key]: expressionEval(expr[key], { rng }) }), {});
    }
    return expr;
  }
  return expr.eval(rng);
};

export const STATUS_NEW = 'new';
export const STATUS_RUNNING = 'running';
export const STATUS_SUSPENDED = 'suspended';
export const STATUS_OK = 'ok';
export const STATUS_FAIL = 'fail';
export const STATUS_STRINGS = [
  'new', // computations have not started
  'running', // computations are in prog
  'suspended', // computations have been suspended, job is not finished
  'ok', // computations are finished, terminated normally
  'fail']; // computations are finished, terminated with error


// -- named constants for job execution pipeline
export const JOB_STATE_NEW = 0;
export const JOB_STATE_RUNNING = 1;
export const JOB_STATE_DONE = 2;
export const JOB_STATE_ERROR = 3;
export const JOB_STATES = [
  JOB_STATE_NEW,
  JOB_STATE_RUNNING,
  JOB_STATE_DONE,
  JOB_STATE_ERROR];


export const TRIAL_KEYS = [
  'tid',
  'result',
  'args',
  'state',
  'book_time',
  'refresh_time',
  'expKey'];

export const range = (start, end) => Array.from({ length: (end - start) }, (v, k) => k + start);

export class Trials {
  constructor(expKey = null, refresh = true) {
    this.ids = [];
    this.dynamicTrials = [];
    this.trials = [];
    this.expKey = expKey;
    if (refresh) {
      this.refresh();
    }
  }

  get length() {
    return this.trials.length;
  }

  refresh = () => {
    if (this.expKey === null) {
      this.trials = this.dynamicTrials
        .filter(trial => trial.state !== JOB_STATE_ERROR);
    } else {
      this.trials = this.dynamicTrials
        .filter(trial => trial.state !== JOB_STATE_ERROR && trial.expKey === this.expKey);
      this.ids = [];
    }
  };

  get tids() {
    return this.trials.map(trial => trial.tid);
  }

  get results() {
    return this.trials.map(trial => trial.result);
  }

  get args() {
    return this.trials.map(trial => trial.args);
  }

  assertValidTrial = (trial) => {
    if (Object.keys(trial).length <= 0) {
      throw new Error('trial should be an object');
    }
    const missingTrialKey = TRIAL_KEYS.find(key => trial[key] === undefined);
    if (missingTrialKey !== undefined) {
      throw new Error(`trial missing key ${missingTrialKey}`);
    }
    if (trial.expKey !== this.expKey) {
      throw new Error(`wrong trial expKey ${trial.expKey}, expected ${this.expKey}`);
    }
    return trial;
  };

  internalIinsertTrialDocs = (docs) => {
    const rval = docs.map(doc => doc.tid);
    this.dynamicTrials = [...this.dynamicTrials, ...docs];
    return rval;
  };

  insertTrialDoc = (trial) => {
    const doc = this.assertValidTrial(trial);
    return this.internalIinsertTrialDocs([doc])[0];
  };

  insertTrialDocs = (trials) => {
    const docs = trials.map(trial => this.assertValidTrial(trial));
    return this.internalIinsertTrialDocs(docs);
  };

  newTrialIds = (N) => {
    const aa = this.ids.length;
    const rval = range(aa, aa + N);
    this.ids = [...this.ids, ...rval];
    return rval;
  };

  newTrialDocs = (tids, specs, results, args) => {
    const rval = [];
    for (let i = 0; i < tids.length; i += 1) {
      const doc = {
        state: JOB_STATE_NEW,
        tid: tids[i],
        spec: specs[i],
        result: results[i],
        args: args[i],
      };
      doc.expKey = this.expKey;
      doc.book_time = null;
      doc.refresh_time = null;
      rval.push(doc);
    }
    return rval;
  };

  deleteAll = () => {
    this.dynamicTrials = [];
    this.refresh();
  };

  countByStateSynced = (arg, trials = null) => {
    const vTrials = trials === null ? this.trials : trials;
    const vArg = Array.isArray(arg) ? arg : [arg];
    const queue = vTrials.filter(doc => vArg.indexOf(doc.state) >= 0);
    return queue.length;
  };

  countByStateUnsynced = (arg) => {
    const expTrials = this.expKey !== null ?
      this.dynamicTrials.map(trial => trial.expKey === this.expKey) : this.dynamicTrials;
    return this.countByStateSynced(arg, expTrials);
  };

  losses = () => this.results.map(r => r.loss);

  statuses = () => this.results.map(r => r.status);

  averageBestError = () => {
    /* const results = this.results.filter(result => result.status === STATUS_OK);
    const loss = results.map(r => r.loss);
    const loss_v = results.map(r => r.loss_variance);
    const true_loss = results.map(r => (r.true_loss === undefined ? r.loss : r.true_loss));

            if (loss_v.every(variance => variance === 0)) {
              return true_loss[argmin(loss)]
            }
            else {
                let cutoff = 0;
                const sigma = Math.sqrt(loss_v[0]);
                while (cutoff < len(loss3) &&loss[cutoff] < loss[0] + 3 * sigma) {
                  cutoff += 1
                }
                const pmin = pmin_sampled(loss3[:cutoff, 0], loss3[:cutoff, 1])
                const avg_true_loss = (pmin * loss3[:cutoff, 2]).sum()
                return avg_true_loss
    */
  };

  get bestTrial() {
    let best = this.trials[0];
    this.trials.forEach((trial) => {
      if (trial.result.status === STATUS_OK && trial.result.loss < best.result.loss) {
        best = trial;
      }
    });
    return best;
  }

  get argmin() {
    const { bestTrial } = this;
    return bestTrial !== undefined ? bestTrial.args : undefined;
  }
}

export class Domain {
  constructor(
    fn, expr,
    {
      name,
    } = {}
  ) {
    this.fn = fn;
    this.expr = expr;
    this.name = name;
  }

  evaluate = async (args) => {
    const rval = await this.fn(args);
    let result;
    if (!Number.isNaN(rval)) {
      result = { loss: rval, status: STATUS_OK };
    } else {
      if (result === undefined) {
        throw new Error('Optimization function should return a loss value');
      }
      const { status, loss } = result;
      if (STATUS_STRINGS.indexOf(status) < 0) {
        throw new Error(`invalid status ${status}`);
      }
      if (status === STATUS_OK && loss === undefined) {
        throw new Error(`invalid loss ${loss}`);
      }
    }
    return result;
  };
  newResult = () => ({
    status: STATUS_NEW,
  });
}
