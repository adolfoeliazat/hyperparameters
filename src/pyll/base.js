class NotImplementedError extends Error {}

export default class BaseSymbol {
  constructor(label, params) {
    this.label = label;
    this.params = params;
  }
  // eslint-disable-next-line no-unused-vars
  eval(rng) {
    // Override this method to generate a new value
    throw new NotImplementedError(this.params);
  }
}
