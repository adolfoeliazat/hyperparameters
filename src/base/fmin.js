/* eslint-disable camelcase */
import RandomState from '../utils/RandomState';
import { Trials, Domain, JOB_STATE_NEW, JOB_STATE_RUNNING, JOB_STATE_ERROR, JOB_STATE_DONE } from './base';

const getTimeStatmp = () => new Date().getTime();

class FMinIter {
  constructor(
    algo, domain, trials,
    {
      rng,
      catch_eval_exceptions = false,
      max_queue_len = 1,
      max_evals = Number.MAX_VALUE,
    } = {}
  ) {
    this.catch_eval_exceptions = catch_eval_exceptions;
    this.algo = algo;
    this.domain = domain;
    this.trials = trials;
    this.max_queue_len = max_queue_len;
    this.max_evals = max_evals;
    this.rng = rng;
  }
  async serial_evaluate(N = -1) {
    let n = N;
    for (let i = 0; i < this.trials.dynamicTrials.length; i += 1) {
      const trial = this.trials.dynamicTrials[i];
      if (trial.state === JOB_STATE_NEW) {
        trial.state = JOB_STATE_RUNNING;
        const now = getTimeStatmp();
        trial.book_time = now;
        trial.refresh_time = now;
        try {
          // eslint-disable-next-line no-await-in-loop
          const result = await this.domain.evaluate(trial.args);
          trial.state = JOB_STATE_DONE;
          trial.result = result;
          trial.refresh_time = getTimeStatmp();
        } catch (e) {
          trial.state = JOB_STATE_ERROR;
          trial.error = `${e}, ${e.message}`;
          trial.refresh_time = getTimeStatmp();
          if (!this.catch_eval_exceptions) {
            this.trials.refresh();
            throw e;
          }
        }
      }
      n -= 1;
      if (n === 0) {
        break;
      }
    }
    this.trials.refresh();
  }

  run = async (N) => {
    const { trials, algo } = this;
    let n_queued = 0;

    const get_queue_len = () => this.trials.countByStateUnsynced(JOB_STATE_NEW);

    let stopped = false;
    while (n_queued < N) {
      let qlen = get_queue_len();
      while (qlen < this.max_queue_len && n_queued < N) {
        const n_to_enqueue = Math.min(this.max_queue_len - qlen, N - n_queued);
        const new_ids = trials.newTrialIds(n_to_enqueue);
        trials.refresh();
        const new_trials = algo(
          new_ids, this.domain, trials,
          this.rng.randrange(0, (2 ** 31) - 1)
        );
        console.assert(new_ids.length >= new_trials.length);
        if (new_trials.length) {
          this.trials.insertTrialDocs(new_trials);
          this.trials.refresh();
          n_queued += new_trials.length;
          qlen = get_queue_len();
        } else {
          stopped = true;
          break;
        }
      }
      // eslint-disable-next-line no-await-in-loop
      await this.serial_evaluate();
      if (stopped) {
        break;
      }
    }
    const qlen = get_queue_len();
    if (qlen) {
      const msg = `Exiting run, not waiting for ${qlen} jobs.`;
      console.error(msg);
    }
  };

  exhaust = async () => {
    const n_done = this.trials.length;
    await this.run(this.max_evals - n_done);
    this.trials.refresh();
    return this;
  }
}

export default async (fn, space, algo, max_evals, params = {}) => {
  const {
    trials: defTrials, rng: rngDefault,
    catch_eval_exceptions = false,
    return_argmin = true
  } = params;

  let rng;
  if (rngDefault) {
    rng = rngDefault;
  } else {
    rng = new RandomState();
  }
  let trials;
  if (!defTrials) {
    trials = new Trials();
  } else {
    trials = defTrials;
  }

  const domain = new Domain(fn, space);

  const rval = new FMinIter(
    algo, domain, trials,
    { max_evals, rng }
  );
  rval.catch_eval_exceptions = catch_eval_exceptions;
  await rval.exhaust();
  if (return_argmin) {
    return trials.argmin;
  }
  return trials;
};
