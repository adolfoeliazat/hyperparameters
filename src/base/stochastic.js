import BaseSymbol, { expressionEval } from './base';

export class Choice extends BaseSymbol {
  eval(rng) {
    const { options } = this.params;
    const idx = rng.randrange(0, options.length, 1);
    const option = options[idx];
    const arg = expressionEval(option, { rng });
    return arg;
  }
}

export class Randint extends BaseSymbol {
  eval(rng) {
    return rng.randrange(0, this.params.upper, 1);
  }
}

export class Uniform extends BaseSymbol {
  eval(rng) {
    const { low, high } = this.params;
    return rng.uniform(low, high);
  }
}

export class QUniform extends BaseSymbol {
  eval(rng) {
    const { low, high, q } = this.params;
    return Math.round(rng.uniform(low, high) / q) * q;
  }
}

export class LogUniform extends BaseSymbol {
  eval(rng) {
    const { low, high } = this.params;
    return Math.exp(rng.uniform(low, high));
  }
}

export class QLogUniform extends BaseSymbol {
  eval(rng) {
    const { low, high, q } = this.params;
    return Math.round(Math.exp(rng.uniform(low, high)) / q) * q;
  }
}

export class Normal extends BaseSymbol {
  eval(rng) {
    const { mu, sigma } = this.params;
    return rng.gauss(mu, sigma);
  }
}

export class QNormal extends BaseSymbol {
  eval(rng) {
    const { mu, sigma, q } = this.params;
    return Math.round(rng.gauss(mu, sigma) / q) * q;
  }
}

export class LogNormal extends BaseSymbol {
  eval(rng) {
    const { mu, sigma } = this.params;
    return Math.exp(rng.gauss(mu, sigma));
  }
}

export class QLogNormal extends BaseSymbol {
  eval(rng) {
    const { mu, sigma, q } = this.params;
    return Math.round(Math.exp(rng.gauss(mu, sigma)) / q) * q;
  }
}


export const sample = (space, params = {}) => {
  const args = expressionEval(space, params);
  if (Object.keys(args).length === 1) {
    const results = Object.keys(args).map(key => args[key]);
    return results.length === 1 ? results[0] : results;
  }
  return args;
};

