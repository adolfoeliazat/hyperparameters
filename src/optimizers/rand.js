import RandomState from '../utils/RandomState';
import { expressionEval } from '../base/base';

// eslint-disable-next-line import/prefer-default-export
export const suggest = (newIds, domain, trials, seed) => {
  const rng = new RandomState(seed);
  let rval = [];
  newIds.forEach((newId) => {
    const paramsEval = expressionEval(domain.expr, { rng });
    const result = domain.newResult();
    rval = [...rval, ...trials.newTrialDocs([newId], [null], [result], [paramsEval])];
  });
  return rval;
};
