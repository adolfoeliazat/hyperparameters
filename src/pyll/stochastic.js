import BaseSymbol from './base';
import RandomState from '../utils/RandomState';

export class Choice extends BaseSymbol {
  eval(rng) {
    const { options } = this.params;
    const idx = rng.randrange(0, options.length, 1);
    const option = options[idx];
    if (!Array.isArray(option)) {
      return option;
    }
    if (option.length !== 2) {
      throw new Error('Array of choice options must consist of label and value');
    }
    const value = option[1];
    return typeof value.eval === 'function' ? value.eval(rng) : value;
  }
}

export class Randint extends BaseSymbol {
  eval(rng) {
    return rng.randrange(0, this.params.upper, 1);
  }
}

export class Uniform extends BaseSymbol {
  eval(rng) {
    const { low, high } = this.params;
    return rng.uniform(low, high);
  }
}

export class QUniform extends BaseSymbol {
  eval(rng) {
    const { low, high, q } = this.params;
    return Math.round(rng.uniform(low, high) / q) * q;
  }
}

export class LogUniform extends BaseSymbol {
  eval(rng) {
    const { low, high } = this.params;
    return Math.exp(rng.uniform(low, high));
  }
}

export class QLogUniform extends BaseSymbol {
  eval(rng) {
    const { low, high, q } = this.params;
    return Math.round(Math.exp(rng.uniform(low, high)) / q) * q;
  }
}

export class Normal extends BaseSymbol {
  eval(rng) {
    const { mu, sigma } = this.params;
    return rng.gauss(mu, sigma);
  }
}

export class QNormal extends BaseSymbol {
  eval(rng) {
    const { mu, sigma, q } = this.params;
    return Math.round(rng.gauss(mu, sigma) / q) * q;
  }
}

export class LogNormal extends BaseSymbol {
  eval(rng) {
    const { mu, sigma } = this.params;
    return Math.exp(rng.gauss(mu, sigma));
  }
}

export class QLogNormal extends BaseSymbol {
  eval(rng) {
    const { mu, sigma, q } = this.params;
    return Math.round(Math.exp(rng.gauss(mu, sigma)) / q) * q;
  }
}


export const sample = (space, params = {}) => {
  let { rng } = params;
  if (!rng) {
    rng = new RandomState();
  }
  return space.eval(rng);
};

