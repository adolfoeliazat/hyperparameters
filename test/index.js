import chai, { assert } from 'chai';
import snapshots from 'chai-snapshot-tests';
import hp, { fmin } from '../src';
import RandomState from '../src/utils/RandomState';
import { sample } from '../src/base/stochastic';
import { suggest as randSuggest } from '../src/optimizers/rand';

chai.use(snapshots(__filename));
const seededSample = (space) => sample(space, { rng: new RandomState(12345) });
const randFMinSeeded = async (opt, space) => await fmin(opt, space, randSuggest, 100, { rng: new RandomState(12345) });


describe('hp.choice.', () => {
  it('is a string', () => {
    const val = seededSample(hp.choice('choice', ['cat', 'dog']));
    assert.typeOf(val, 'string');
  });
  it('is one of the elements', () => {
    const val = seededSample(hp.choice('choice', ['cat', 'dog']));
    assert(['cat', 'dog'].indexOf(val) >= 0, 'val was actually ' + val);
  });
  it('picks a number', () => {
    const val = seededSample(hp.choice('numbers', [1, 2, 3, 4]));
    assert(val === 4, 'val was actually: ' + val);
  });
});

describe('hp.randint.', () => {
  it('in range [0,5)', () => {
    const val  = seededSample(hp.randint('randint', 5));
    assert(val >= 0 && val < 5, `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('upper: negative', seededSample(hp.randint('randint', -2)));
    assert.snapshot('upper: 0', seededSample(hp.randint('randint', 0)));
    assert.snapshot('upper: 1', seededSample(hp.randint('randint', 1)));
    assert.snapshot('upper: 1000000', seededSample(hp.randint('randint', 1000000)));
  });

});

describe('hp.uniform.', () => {
  it('between 0 and 1', () => {
    const val  = seededSample(hp.uniform('uniform', 0, 1));
    assert(val >= 0 && val <= 1, `actual value ${val}`);
  });
  it('between -1 and 1', () => {
    const val  = seededSample(hp.uniform('uniform', -1, 1));
    assert(val >= -1 && val <= 1, `actual value ${val}`);
  });

  it('Snapshot tests', () => {
    assert.snapshot('uniform -1, 1', seededSample(hp.uniform('uniform', -1, 1)));
    assert.snapshot('uniform -100000, -1', seededSample(hp.uniform('uniform', -100000, -1)));
    assert.snapshot('uniform -1, -10', seededSample(hp.uniform('uniform', -1, -10)));
    assert.snapshot('uniform 5, 1', seededSample(hp.uniform('uniform', 5, 1)));
    assert.snapshot('uniform 1, 1000000', seededSample(hp.uniform('uniform', 1, 1000000)));
    assert.snapshot('uniform 1, 1', seededSample(hp.uniform('uniform', 1, 1)));
  });
});

describe('hp.quniform.', () => {
  it('between 0 and 1', () => {
    const val  = seededSample(hp.quniform('quniform', 0, 1, 0.2));
    assert(val >= 0 && val <= 1, `actual value ${val}`);
  });
  it('between -1 and 1, step 1', () => {
    const val  = seededSample(hp.quniform('quniform', -1, 1, 1));
    assert(val === -1 || val === 0 || val === 1, `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('quniform -1, 1, 0.1', seededSample(hp.quniform('quniform', -1, 1, 0.1)));
    assert.snapshot('quniform -100000, -1, -1', seededSample(hp.quniform('quniform', -100000, -1, -1)));
    assert.snapshot('quniform -1, -10, 0.22222', seededSample(hp.quniform('quniform', -1, -10, 0.22222)));
    assert.snapshot('quniform 5, 1, -0.111', seededSample(hp.quniform('quniform', 5, 1, -0.111)));
    assert.snapshot('quniform 1, 1000000, 50', seededSample(hp.quniform('quniform', 1, 1000000, 50)));
    assert.snapshot('quniform 1, 1, 0.001', seededSample(hp.quniform('quniform', 1, 1, 0.001)));
  });
});

describe('hp.loguniform.', () => {
  it('between e^0 and e^1', () => {
    const low = 0;
    const high = 1;
    const val  = seededSample(hp.loguniform('loguniform', low, high));
    assert(val >= Math.exp(low) && val <= Math.exp(high), `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('loguniform -1, 1', seededSample(hp.loguniform('loguniform', -1, 1)));
    assert.snapshot('loguniform -100000, -1', seededSample(hp.loguniform('loguniform', -100000, -1)));
    assert.snapshot('loguniform -1, -10', seededSample(hp.loguniform('loguniform', -1, -10)));
    assert.snapshot('loguniform 5, 1', seededSample(hp.loguniform('loguniform', 5, 1)));
    assert.snapshot('loguniform 100, 100', seededSample(hp.loguniform('loguniform', 100, 100)));
    assert.snapshot('loguniform 1, 1', seededSample(hp.loguniform('loguniform', 1, 1)));
  });
});

describe('hp.qloguniform.', () => {
  it('e^0 and e^1', () => {
    const low = 0;
    const high = 1;
    const val  = seededSample(hp.qloguniform('qloguniform', low, high, 0.2));
    assert(val >= Math.exp(low) && val <= Math.exp(high), `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('qloguniform -1, 1, 0.1', seededSample(hp.qloguniform('qloguniform', -1, 1, 0.1)));
    assert.snapshot('qloguniform -100000, -1, -1', seededSample(hp.qloguniform('qloguniform', -100000, -1, -1)));
    assert.snapshot('qloguniform -1, -10, 0.2222', seededSample(hp.qloguniform('qloguniform', -1, -10, 0.22222)));
    assert.snapshot('qloguniform 5, 1, -0.111', seededSample(hp.qloguniform('qloguniform', 5, 1, -0.111)));
    assert.snapshot('qloguniform 100, 50, 50', seededSample(hp.qloguniform('qloguniform', 100, 50, 50)));
    assert.snapshot('qloguniform 1, 1, 0.001', seededSample(hp.qloguniform('qloguniform', 1, 1, 0.001)));
  });
});

describe('hp.normal.', () => {
  it('a number', () => {
    const mu = -1;
    const sigma = 1;
    const val  = seededSample(hp.normal('normal', mu, sigma));
    assert(!isNaN(val), `actual value ${val}`);
  });
  it('within 3 standard deviations of mean', () => {
    const mu = 0;
    const sigma = 1;
    const val  = seededSample(hp.normal('normal', mu, sigma));
    assert(val >= mu - (3*sigma) && val <= mu + (3*sigma), `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('normal -1, 1', seededSample(hp.normal('normal', -1, 1)));
    assert.snapshot('normal -100000, -1', seededSample(hp.normal('normal', -100000, -1)));
    assert.snapshot('normal -1, -10', seededSample(hp.normal('normal', -1, -10)));
    assert.snapshot('normal 5, 1', seededSample(hp.normal('normal', 5, 1)));
    assert.snapshot('normal 100, 100', seededSample(hp.normal('normal', 100, 100)));
    assert.snapshot('normal 1, 1', seededSample(hp.normal('normal', 1, 1)));
  });
});

describe('hp.qnormal.', () => {
  it('a number', () => {
    const mu = -1;
    const sigma = 1;
    const val  = seededSample(hp.normal('normal', mu, sigma));
    assert(!isNaN(val), `actual value ${val}`);
  });
  it('within 3 standard deviations of mean', () => {
    const mu = 0;
    const sigma = 1;
    const val  = seededSample(hp.qnormal('qnormal', mu, sigma, 0.1));
    assert(val >= mu - (3*sigma) && val <= mu + (3*sigma), `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('qnormal -1, 1, 0.1', seededSample(hp.qnormal('qnormal', -1, 1, 0.1)));
    assert.snapshot('qnormal -100000, -1, -1', seededSample(hp.qnormal('qnormal', -100000, -1, -1)));
    assert.snapshot('qnormal -1, -10, 0.22222', seededSample(hp.qnormal('qnormal', -1, -10, 0.22222)));
    assert.snapshot('qnormal 5, 1, -0.111', seededSample(hp.qnormal('qnormal', 5, 1, -0.111)));
    assert.snapshot('qnormal 1, 1000000, 50', seededSample(hp.qnormal('qnormal', 1, 1000000, 50)));
    assert.snapshot('qnormal 1, 1, 0.001', seededSample(hp.qnormal('qnormal', 1, 1, 0.001)));
  });
});

describe('hp.lognormal.', () => {
  it('positive', () => {
    const mu = 0;
    const sigma = 1;
    const val  = seededSample(hp.lognormal('lognormal', mu, sigma));
    assert(val >= 0, `actual value ${val}`);
  });
  it('less ~e^3 from the mean, or less than ~3 standard deviations from it', () => {
    const mu = 0;
    const sigma = 1;
    const val  = seededSample(hp.lognormal('lognormal', mu, sigma));
    assert(val <= 50, `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('lognormal -1, 1', seededSample(hp.lognormal('lognormal', -1, 1)));
    assert.snapshot('lognormal -100000, -1', seededSample(hp.lognormal('lognormal', -100000, -1)));
    assert.snapshot('lognormal -1, -10', seededSample(hp.lognormal('lognormal', -1, -10)));
    assert.snapshot('lognormal 5, 1', seededSample(hp.lognormal('lognormal', 5, 1)));
    assert.snapshot('lognormal 100, 100', seededSample(hp.lognormal('lognormal', 100, 100)));
    assert.snapshot('lognormal 1, 1', seededSample(hp.lognormal('lognormal', 1, 1)));
  });
});

describe('hp.qlognormal.', () => {
  it('a number', () => {
    const mu = -1;
    const sigma = 1;
    const val  = seededSample(hp.qlognormal('qlognormal', mu, sigma, 0.2));
    assert(!isNaN(val), `actual value ${val}`);
  });
  it('within 3 standard deviations of mean', () => {
    const mu = 0;
    const sigma = 1;
    const val  = seededSample(hp.qlognormal('qlognormal', mu, sigma, 0.1));
    assert(val >= mu - (3*sigma) && val <= mu + (3*sigma), `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('qlognormal -1, 1, 0.1', seededSample(hp.qlognormal('qlognormal', -1, 1, 0.1)));
    assert.snapshot('qlognormal -100000, -1, -1', seededSample(hp.qlognormal('qlognormal', -100000, -1, -1)));
    assert.snapshot('qlognormal -1, -10, 0.22222', seededSample(hp.qlognormal('qlognormal', -1, -10, 0.22222)));
    assert.snapshot('qlognormal 5, 1, -0.111', seededSample(hp.qlognormal('qlognormal', 5, 1, -0.111)));
    assert.snapshot('qlognormal 100, 100, 50', seededSample(hp.qlognormal('qlognormal', 100, 100, 50)));
    assert.snapshot('qlognormal 1, 1, 0.001', seededSample(hp.qlognormal('qlognormal', 1, 1, 0.001)));
  });
});

describe('sample', () => {
  it('Choice as array', () => {
    const space = hp.choice('a',
      [
        hp.lognormal('c1', 0, 1),
        hp.uniform('c2', -10, 10)
      ]
    );

    assert.snapshot('sample: array', seededSample(space));
  });
  it('more complex space with depth', () => {
    const space = {
      x: hp.normal('x', 0, 2),
      y: hp.uniform('y', 0, 1),
      choice: hp.choice('choice', [
        null, hp.uniform('float', 0, 1),
      ]),
      array: [
        hp.normal('a', 0, 2), hp.uniform('b', 0, 3), hp.choice('c', [false, true]),
      ],
      obj: {
        u: hp.uniform('u', 0, 3),
        v: hp.uniform('v', 0, 3),
        w: hp.uniform('w', -3, 0)
      }
    };
    assert.snapshot('sample: depth', seededSample(space));
  });
});
describe('fmin + rand', () => {
  it('FMin for x^2 - x + 1', async () => {
    const space = hp.uniform('x', -5, 5);
    const opt = x => ((x ** 2) - (x + 1));
    assert.snapshot('FMin for x^2 - x + 1', await randFMinSeeded(opt, space));
  });
  it('Hyperparameters space', async () => {
    const space = {
      x: hp.uniform('x', -5, 5),
      y: hp.uniform('y', -5, 5)
    };
    const opt = ({ x, y }) => ((x ** 2) + (y ** 2));
    assert.snapshot('Hyperparameters space', await randFMinSeeded(opt, space));
  });
  it('Choice selection of expressions', async () => {
    const space = hp.choice('a',
      [
        hp.lognormal('c1', 0, 1),
        hp.uniform('c2', -10, 10)
      ]
    );
    const opt = ( x ) => (x ** 2);
    assert.snapshot('choice as array space', await randFMinSeeded(opt, space));
  });
  it('Deep learning space', async () => {
    const space = {
      // Learning rate should be between 0.00001 and 1
      learning_rate:
        hp.loguniform('learning_rate', Math.log(1e-5), Math.log(1)),
      use_double_q_learning:
        hp.choice('use_double_q_learning', [true, false]),
      layer1_size: hp.quniform('layer1_size', 10, 100, 1),
      layer2_size: hp.quniform('layer2_size', 10, 100, 1),
      layer3_size: hp.quniform('layer3_size', 10, 100, 1),
      future_discount_max: hp.uniform('future_discount_max', 0.5, 0.99),
      future_discount_increment:
        hp.loguniform('future_discount_increment', Math.log(0.001), Math.log(0.1)),
      recall_memory_size: hp.quniform('recall_memory_size', 1, 100, 1),
      recall_memory_num_experiences_per_recall:
        hp.quniform('recall_memory_num_experiences_per_recall', 10, 2000, 1),
      num_epochs: hp.quniform('num_epochs', 1, 10, 1),
    };

    const opt = params => params.learning_rate ** 2;

    assert.snapshot('Deep learning space', await randFMinSeeded(opt, space));
  });
});

