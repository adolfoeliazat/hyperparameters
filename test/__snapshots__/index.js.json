{
  "upper: negative": 0,
  "upper: 0": 0,
  "upper: 1": 0,
  "upper: 1000000": 933394,
  "uniform -1, 1": 0.7803094370686219,
  "uniform -100000, -1": -10985.418301287442,
  "uniform -1, -10": -9.011392466808799,
  "uniform 5, 1": 1.4393811258627562,
  "uniform 1, 1000000": 890154.8283795924,
  "uniform 1, 1": 1,
  "quniform -1, 1, 0.1": 0.8,
  "quniform -100000, -1, -1": -10985,
  "quniform -1, -10, 0.22222": -9.11102,
  "quniform 5, 1, -0.111": 1.443,
  "quniform 1, 1000000, 50": 890150,
  "quniform 1, 1, 0.001": 1,
  "loguniform -1, 1": 2.1821473983314923,
  "loguniform -100000, -1": 0,
  "loguniform -1, -10": 0.00012201184023802779,
  "loguniform 5, 1": 4.218084545621744,
  "loguniform 100, 100": 2.6881171418161356e+43,
  "loguniform 1, 1": 2.718281828459045,
  "qloguniform -1, 1, 0.1": 2.2,
  "qloguniform -100000, -1, -1": 0,
  "qloguniform -1, -10, 0.2222": 0,
  "qloguniform 5, 1, -0.111": 4.218,
  "qloguniform 100, 50, 50": 1.258879250524572e+24,
  "qloguniform 1, 1, 0.001": 2.718,
  "normal -1, 1": -0.3037819091033397,
  "normal -100000, -1": -100000.6962180909,
  "normal -1, -10": -7.962180908966603,
  "normal 5, 1": 5.69621809089666,
  "normal 100, 100": 169.62180908966604,
  "normal 1, 1": 1.6962180908966604,
  "qnormal -1, 1, 0.1": -0.30000000000000004,
  "qnormal -100000, -1, -1": -100001,
  "qnormal -1, -10, 0.22222": -7.99992,
  "qnormal 5, 1, -0.111": 5.6610000000000005,
  "qnormal 1, 1000000, 50": 696200,
  "qnormal 1, 1, 0.001": 1.696,
  "lognormal -1, 1": 0.7380218047375157,
  "lognormal -100000, -1": 0,
  "lognormal -1, -10": 0.00034839247604157605,
  "lognormal 5, 1": 297.7392462565868,
  "lognormal 100, 100": 4.632502881060729e+73,
  "lognormal 1, 1": 5.453284517439546,
  "qlognormal -1, 1, 0.1": 0.7000000000000001,
  "qlognormal -100000, -1, -1": 0,
  "qlognormal -1, -10, 0.22222": 0,
  "qlognormal 5, 1, -0.111": 297.702,
  "qlognormal 100, 100, 50": 4.632502881060729e+73,
  "qlognormal 1, 1, 0.001": 5.453,
  "sample: array": -3.6724888806975553,
  "sample: depth": {
    "x": 1.3924361817933206,
    "y": 0.13070729472270082,
    "choice": null,
    "array": [
      -0.6553433999116457,
      0.11927848398147867,
      false
    ],
    "obj": {
      "u": 2.47930839269838,
      "v": 1.703175088041835,
      "w": -1.4037662402075997
    }
  },
  "FMin for x^2 - x + 1": {},
  "Hyperparameters space": {
    "x": "-0.7975415",
    "y": "-0.7414791"
  },
  "choice as array space": {},
  "Deep learning space": {
    "learning_rate": "0.0000123",
    "use_double_q_learning": true,
    "layer1_size": "29.0000000",
    "layer2_size": "41.0000000",
    "layer3_size": "56.0000000",
    "future_discount_max": "0.6085554",
    "future_discount_increment": "0.0013308",
    "recall_memory_size": "8.0000000",
    "recall_memory_num_experiences_per_recall": "1275.0000000",
    "num_epochs": "2.0000000"
  }
}